src/stats.ts

```diff
@@ -0,0 +1,121 @@
+import { isDfid } from './util/is';
+
+export class StatsClient extends Client {
+  /**
+   * Registers a new session with Doofinder, and returns
+   * the session ID generated for this session
+   *
+   * @param   {String}     sessionId   The sessionId
+   *
+   * @param   {Function}   callback    Optional, callback after everything works
+   *
+   * @return  {String}     The newly generated Session ID
+   */
+  public async requestSession(sessionId: string, hashid?: string): Promise<GenericObject> {
+    if (hashid) {
+      this.hashid = hashid;
+    }
+    return await this.stats(StatsEvent.Init, { session_id: sessionId, hashid: this.hashid });
+  }
+
+  /**
+   * Registers a click within this session, using the dfid
+   * or the database id and datatype.
+   *
+   * @param  {String}   sessionId           The sessionId
+   *
+   * @param  {String}   id                  Id of the result or Doofinder's internal ID
+   *                                        for the result.
+   *
+   * @param  {String}   datatype            Optional. If the id is not a Doofinder id
+   *                                        this argument is required.
+   *
+   * @param  {String}   query               Optional. Search terms.
+   *
+   * @param  {String}   custom_results_id   Optional. Id of the custom results
+   *                                        that produced the current set of
+   *                                        results, including the current one
+   *                                        being clicked.
+   *
+   */
+  public async registerClick(
+    sessionId: string,
+    id: string,
+    datatype?: string,
+    query?: string,
+    customResultsId?: string | number
+  ): Promise<GenericObject> {
+    const params: GenericObject = {
+      session_id: sessionId,
+      hashid: this.hashid,
+    };
+
+    if (isDfid(id)) {
+      params['dfid'] = id;
+    } else if (datatype) {
+      params['id'] = id;
+      params['datatype'] = datatype;
+    } else {
+      throw new Error('Must declare a dfid or an id and datatype');
+    }
+
+    params['query'] = query || '';
+
+    if (customResultsId) {
+      params['custom_results_id'] = customResultsId;
+    }
+
+    return await this.stats(StatsEvent.Click, params);
+  }
+
+  /**
+   * Register a checkout within the session
+   *
+   * @param   {String}    sessionId   Optional. The session ID
+   *
+   */
+  public async registerCheckout(sessionId: string): Promise<GenericObject> {
+    const params = {
+      session_id: sessionId,
+      hashid: this.hashid,
+    };
+
+    return await this.stats(StatsEvent.Checkout, params);
+  }
+
+  /**
+   *
+   * Register the display banner events
+   *
+   * @param  {Number}    bannerId    The banner ID
+   *
+   */
+  public async registerDisplayBannerEvent(sessionId: string, bannerId: number): Promise<GenericObject> {
+    const params = {
+      session_id: sessionId,
+      hashid: this.hashid,
+      banner_id: '' + bannerId,
```

Comment:
This conversion shouldn't be necessary.

Reply:
Typescript does think on the contrary.

Reply:
What if you tell TS that bannerId is a string? Indeed it's not going used in arithmetics but as a string value in the query string of the request, so it's not wrong in my opinion.

Reply:
I'm thinking about future users here. If we force string, we are limiting the options to call the function to third parties.

Reply:
I don't get that, you're converting the number to string… but you can't pass a string because you said it must be a number… A number can be converted to a string but the opposite is not true, so I think having the parameter as a number is much more limiting than converting it to string.